### PR TITLE
Feature/connection options

### PR DESCRIPTION
**Pull request by koden-km**
I'm not sure what should be done for the error message in HandshakeController::onTimeout().
The timeout can be null now and is by default.

```
'the handshake timed out after ' . $this->options->connectionTimeout() . ' seconds'
```

### CODE REVIEW

**Comment by jmalloc** [comment]
Just change it to `"the handshake timed out"`.

File: src/ConnectionOptions.php [hunk 1]
```diff
@@ -101,17 +120,84 @@ public function productName()
...
+     *
+     * @return integer|float|null The timeout, in seconds.
+     */
+    public function heartbeatTimeout()
+    {
+        return $this->heartbeatTimeout;
+    }
+
+    /**
+     * Set how often the server and client must send heartbeat packets to keep the connection alive.
```
Comment (jmalloc) [inline src/ConnectionOptions.php]: s/packet/frame

File: src/ConnectionOptions.php [hunk 2]
```diff
@@ -101,17 +120,84 @@ public function productName()
...
+        $options = clone $this;
+        $options->connectionTimeout = $timeout;
+
+        return $options;
+    }
+
+    /**
+     * Get how often the server and client must send heartbeat packets to keep the connection alive.
+     *
+     * @return integer|float|null The timeout, in seconds.
```
Comment (jmalloc) [inline src/ConnectionOptions.php]: Need to explain what `null` means for all these timeouts.

**Comment by jmalloc** [comment]
Is this ready to go?

**Comment by koden-km** [comment]
Yeah, i think i got everything this time. Look it over to check.
I considered renaming the fallback connection timeout from DEFAULT_SOCKET_TIMEOUT to FALLBACK_SOCKET_TIMEOUT because thats kind of what it is.

File: src-internal/v091/Transport/HandshakeController.php [hunk 3]
```diff
@@ -223,10 +223,14 @@ private function onTuneFrame(ConnectionTuneFrame $frame)
             $this->handshakeResult->maximumFrameSize = $frame->frameMax;
         }
 
-        // @todo Use heartbeat value from connection options
-        // @link https://github.com/recoilphp/amqp/issues/1
+        $optionsHeartbeat = $this->options->heartbeatInterval();
```
Comment (jmalloc) [inline src-internal/v091/Transport/HandshakeController.php]: Forgive me for being anal with micro-optimisations in any of the transport/protocol code, but we should avoid this call to heartbeatInterval() if the server is using `HEARTBEAT_DISABLED`.

File: src/ConnectionOptions.php [hunk 4]
```diff
@@ -101,17 +120,84 @@ public function productName()
...
+        $options = clone $this;
+        $options->connectionTimeout = $timeout;
+
+        return $options;
+    }
+
+    /**
+     * Get how often the server and client must send heartbeat frames to keep the connection alive.
+     *
+     * @return integer|float|null The interval, in seconds.
```
Comment (jmalloc) [inline src/ConnectionOptions.php]: Need to explain `null` here too. "The heatbeat interval (in seconds), or null to use the interval suggested by the server."

**Comment by jmalloc** [comment]
A few minor things, otherwise looks great!

**Pull request merged**